File: luda/editor/client/src/timeline/PlaybackTimeView.ts
import {
  BorderPosition,
  ColorUtil,
  FontWeight,
  Language,
  Rect,
  Render,
  Text,
  TextAlign,
  TextBaseline,
  Translate,
  XywhRect,
} from "namui";

export const PlaybackTimeView: Render<
  {},
  {
    playbackTimeMs: number;
    layout: XywhRect;
  }
> = (state, props) => {
  const minutes = Math.floor(props.playbackTimeMs / 60000);
  const seconds = Math.floor((props.playbackTimeMs % 60000) / 1000);
  const milliseconds = Math.floor(props.playbackTimeMs % 1000);

  const MM = minutes.toString().padStart(2, "0");
  const ss = seconds.toString().padStart(2, "0");
  const mmm = milliseconds.toString().padStart(3, "0");
  return [
    Rect({
      ...props.layout,
      style: {
        stroke: {
          borderPosition: BorderPosition.inside,
          color: ColorUtil.Black,
          width: 1,
        },
      },
    }),
    Text({
      x: props.layout.x + props.layout.width / 2,
      y: props.layout.y + props.layout.height / 2,
      align: TextAlign.center,
      baseline: TextBaseline.middle,
      fontType: {
        fontWeight: FontWeight.regular,
        language: Language.ko,
        serif: false,
        size: 20,
      },
      style: {
        color: ColorUtil.Black,
      },
      text: `${MM}:${ss}:${mmm}`,
    }),
  ];
};

File: luda-editor-client/src/editor/timeline/playback_time_view.rs
use ::namui::*;
use chrono::Duration;

pub struct PlaybackTimeView {}
pub struct PlaybackTimeViewProps {
    pub playback_time: Duration,
    pub xywh: namui::XywhRect<f32>,
}

impl PlaybackTimeView {
    pub fn new() -> Self {
        PlaybackTimeView {}
    }
}

impl namui::Entity for PlaybackTimeView {
    type Props = PlaybackTimeViewProps;
    fn update(&mut self, event: &dyn std::any::Any) {}
    fn render(&self, props: &Self::Props) -> namui::RenderingTree {
        let num_seconds = props.playback_time.num_seconds();

        let seconds = num_seconds % 60;
        let minutes = (num_seconds / 60) % 60;
        let milliseconds = props.playback_time.num_milliseconds() % 1000;

        let text = format!("{:02}:{:02}.{:03}", minutes, seconds, milliseconds);

        render![
            namui::rect(namui::RectParam {
                x: props.xywh.x,
                y: props.xywh.y,
                width: props.xywh.width,
                height: props.xywh.height,
                style: namui::RectStyle {
                    stroke: Some(namui::RectStroke {
                        border_position: namui::BorderPosition::Inside,
                        color: namui::Color::BLACK,
                        width: 1.0,
                    }),
                    ..Default::default()
                },
                ..Default::default()
            }),
            namui::text(namui::TextParam {
                x: props.xywh.x + props.xywh.width / 2.0,
                y: props.xywh.y + props.xywh.height / 2.0,
                align: namui::TextAlign::Center,
                baseline: namui::TextBaseline::Middle,
                font_type: namui::FontType {
                    font_weight: namui::FontWeight::REGULAR,
                    language: namui::Language::Ko,
                    serif: false,
                    size: 20,
                },
                style: namui::TextStyle {
                    color: namui::Color::BLACK,
                    ..Default::default()
                },
                text: text,
            })
        ]
    }
}
